packages/alfa-result/src/result.ts

```diff
@@ -32,9 +32,10 @@ export interface Result<T, E = T>
   map<U>(mapper: Mapper<T, U>): Result<U, E>;
   mapErr<F>(mapper: Mapper<E, F>): Result<T, F>;
   mapOrElse<U>(ok: Mapper<T, U>, err: Mapper<E, U>): U;
+  apply<U>(mapper: Result<Mapper<T, U>, E>): Result<U, E>;
   flatMap<U>(mapper: Mapper<T, Result<U, E>>): Result<U, E>;
+  flatten<T, E>(this: Result<Result<T, E>, E>): Result<T, E>;
```

Comment:
Could be
```suggestion
  flatten<T, E, F>(this: Result<Result<T, E>, F>): Result<T, E | F>;
```
to be a bit more generic.

Also, it is a bit confusing to re-use the names `T` and `E` which are already the global type parameter. All other signatures are using `U` or `F` when they need new parameters.

Reply:
"Fresh" type parameters are needed for properly typing `this` and so the names should align with the ones used for the class itself.

I'm not too keen on further complicating the generics for no current benefit so let's keep it simple 🙈